tests/unit/stage/test_serialize_pipeline_file.py

```diff
@@ -0,0 +1,196 @@
+
+import pytest
+from voluptuous import Schema as _Schema
+
+from dvc import output
+from dvc.dvcfile import PIPELINE_FILE
+from dvc.schema import SINGLE_PIPELINE_STAGE_SCHEMA
+from dvc.stage import PipelineStage, create_stage
+from dvc.stage.serialize import to_pipeline_file as _to_pipeline_file
+
+kwargs = {"name": "something", "cmd": "command", "path": PIPELINE_FILE}
+Schema = _Schema(SINGLE_PIPELINE_STAGE_SCHEMA)
+
+
+def to_pipeline_file(stage):
+    """Validate schema on each serialization."""
+    e = _to_pipeline_file(stage)
+    assert len(Schema(e)) == 1
+    return e
+
+
+def test_cmd(dvc):
+    stage = create_stage(PipelineStage, dvc, **kwargs)
+    entry = to_pipeline_file(stage)
+    assert entry == {"something": {"cmd": "command"}}
+
+
+def test_wdir(dvc):
+    stage = create_stage(PipelineStage, dvc, **kwargs)
+    assert stage.PARAM_WDIR not in to_pipeline_file(stage)["something"]
+
+    stage.wdir = os.curdir
+    assert stage.PARAM_WDIR not in to_pipeline_file(stage)["something"]
+
+    stage.wdir = "some-dir"
+    assert to_pipeline_file(stage)["something"][stage.PARAM_WDIR] == "some-dir"
+
+
+def test_deps_sorted(dvc):
+    stage = create_stage(
+        PipelineStage, dvc, deps=["a", "quick", "lazy", "fox"], **kwargs
+    )
+    assert to_pipeline_file(stage)["something"][stage.PARAM_DEPS] == [
+        "a",
+        "fox",
+        "lazy",
+        "quick",
+    ]
+
+
+def test_outs_sorted(dvc):
+    stage = create_stage(
+        PipelineStage,
+        dvc,
+        outs=["too", "many", "outs"],
+        deps=["foo"],
+        **kwargs,
+    )
+    assert to_pipeline_file(stage)["something"][stage.PARAM_OUTS] == [
+        "many",
+        "outs",
+        "too",
+    ]
+
+
+def test_params_sorted(dvc):
+    params = [
+        "lorem",
+        "ipsum",
+        {"custom.yaml": ["wxyz", "pqrs", "baz"]},
+        {"params.yaml": ["barr"]},
+    ]
+    stage = create_stage(
+        PipelineStage, dvc, outs=["bar"], deps=["foo"], params=params, **kwargs
+    )
+    assert to_pipeline_file(stage)["something"][stage.PARAM_PARAMS] == [
+        "barr",
+        "ipsum",
+        "lorem",
+        {"custom.yaml": ["baz", "pqrs", "wxyz"]},
+    ]
+
+
+def test_params_file_sorted(dvc):
+    params = [
+        "lorem",
+        "ipsum",
+        {"custom.yaml": ["wxyz", "pqrs", "baz"]},
+        {"a-file-of-params.yaml": ["barr"]},
+    ]
+    stage = create_stage(
+        PipelineStage, dvc, outs=["bar"], deps=["foo"], params=params, **kwargs
+    )
+    assert to_pipeline_file(stage)["something"][stage.PARAM_PARAMS] == [
+        "ipsum",
+        "lorem",
+        {"a-file-of-params.yaml": ["barr"]},
+        {"custom.yaml": ["baz", "pqrs", "wxyz"]},
+    ]
+
+
+@pytest.mark.parametrize(
+    "typ, extra",
+    [("plots", {"plot": True}), ("metrics", {"metric": True}), ("outs", {})],
+)
+def test_outs_and_outs_flags_are_sorted(dvc, typ, extra):
+    stage = create_stage(PipelineStage, dvc, deps=["input"], **kwargs)
```

Comment:
Maybe, we should implement `{metrics,plots}[_persist][_no_cache]` that are missing?

Reply:
@skshetry Could you elaborate?

Reply:
On `.dvc` and `dvc.yaml` files, we allow metrics and plots to be `cache: False` and `persist: True`. But, it cannot be set from outside (eg: `fill_stage_outputs` and `dvc run`).  So, I was just saying that maybe we should implement that for `run`.

Reply:
@skshetry Let's not create those `dvc run` options for now, it is crowded there as it is. We clearly need to consider a different CLI approach, as `metrics_persist_no_cache` is just too ugly and doesn't scale if we introduce yet another option :slightly_smiling_face: We've been discussing something like using a comma to join those fstab-style, but, again, not worth bothering with that for now.